### PR TITLE
Issue 2

### PR DESCRIPTION
**Pull request by Nathmel123**
Implemented mei-xml conversion.

### CODE REVIEW

File: Classes/Common/XmlDocument.php [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+        $result = [];
+        // Parse attributes
+        $attributes = $node->attributes();
+        foreach ($attributes as $attrName => $attrValue) {
+            $trimmedValue = trim(strval($attrValue));
+            if (!empty($trimmedValue)) {
+                $result['@attributes'][$attrName] = $trimmedValue;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (dikastes) [reply Classes/Common/XmlDocument.php:84]: See https://laravel.com/docs/11.x/collections#method-mapwithkeys

File: Classes/Common/XmlDocument.php [hunk 2]
```diff
@@ -0,0 +1,143 @@
+<?php
+
+namespace Slub\LisztCommon\Common;
+use SimpleXMLElement;
```
Comment (dikastes) [inline Classes/Common/XmlDocument.php:4]: use Illuminate\Support\Collection;

Comment (dikastes) [reply Classes/Common/XmlDocument.php:4]: see below why

File: Classes/Common/XmlDocument.php [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+            $result['@value'] = $nodeValue;
+        }
+
+        // Include xml:id attribute
+        if($this->include_xml_id) {
+
+            $xmlId = $node->attributes('xml', true)->id;
+            $trimmedXmlId = trim(strval($xmlId));
+            if (!empty($trimmedXmlId)) {
+            $result['@xml:id'] = $trimmedXmlId;
```
Comment (dikastes) [inline Classes/Common/XmlDocument.php:98]: indentation

File: Classes/Common/XmlDocument.php [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            }
+
+        }
+
```
Comment (dikastes) [inline Classes/Common/XmlDocument.php:102]: superfluous empty line

**Review by dikastes — CHANGES_REQUESTED** [review]
Great result, some minor formatting issues and I have some small and two big desires concerning styling.

**Comment by Nathmel123** [comment]
Collection usage for children is not functional, please do not merge.

**Comment by Nathmel123** [comment]
Illuminate\Support\Collection->each() does not work properly. I replaced it with foreach, maybe there is another solution, but for the moment this is the best solution.

File: Classes/Common/XmlDocument.php [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            if($found) {
+                return $result;
+            }
+            $childData = $this->__convert($childNode);
+            // Always parse child nodes as array
+            if (!isset($result[$childName])) {
+                $result[$childName] = [];
+            }
+            $result[$childName][] = $childData;
+        }
```
[the remark this answers is not in the document]
Comment (Nathmel123) [reply Classes/Common/XmlDocument.php:142]: See my comment. The foreach loop is the best solution for now, Collection->each() does weird things.

**Comment by Nathmel123** [comment]
If a node is splited, its xmlId is not included, if xmlId parsing is off. It should be enabled.

File: Classes/Common/XmlDocument.php [hunk 6]
```diff
@@ -2,30 +2,175 @@
...
+    public static function from(string $xmlString): XmlDocument
     {
         return new XmlDocument($xmlString);
     }
 
     public function toArray(): array
     {
-        // add function here
-        return [];
+        // Check if array is already converted
```
Comment (dikastes) [inline Classes/Common/XmlDocument.php:69]: great idea!

File: Classes/Common/XmlDocument.php [hunk 7]
```diff
@@ -2,30 +2,175 @@
...
 
-    public function toJson(): string
+
+    public function toJson(): array
     {
-        // add function here
-        return '';
+        $result = [];
+        $xmlArray = $this->toArray();
+        foreach ($xmlArray as $id => $value) {
```
Comment (dikastes) [inline Classes/Common/XmlDocument.php:85]: ? ```php $this->convertedArray``` should always consist of just one element?

File: Classes/Common/XmlDocument.php [hunk 8]
```diff
@@ -2,30 +2,175 @@
...
+            $result[$id] = json_encode($value, JSON_PRETTY_PRINT);
+        }
+        return $result;
```
Comment (dikastes) [inline Classes/Common/XmlDocument.php:88]: I'd expect a JSON string, not an array

File: Tests/Unit/Common/XmlDocumentTest.php [hunk 9]
```diff
@@ -11,27 +11,50 @@
...
     }
+
+    /**
+     * @test
+     */
+
+    public function xmlStringEqualFile() {
+        self::assertXmlStringEqualsXmlFile('Tests/Testfiles/meitest2.xml',$this->xmlString);
+    }
+
```
Comment (dikastes) [inline Tests/Unit/Common/XmlDocumentTest.php:58]: provide more tests (probably not making use of files) with minimal examples:
- convert attribute
- convert text value
- convert mixed content
- all config options

**Review by dikastes — CHANGES_REQUESTED** [review]

**Comment by Nathmel123** [comment]
Somehow the checks fail at my machine, if I merge dev into issue-2 and than pull

**Comment by Nathmel123** [comment]
Fixed

**Review by dikastes — APPROVED** [review]

**Pull request merged**